File: engine-cpp/src/orderbook/Trade.cpp
#include "orderbook/Trade.hpp"

#include <stdexcept>
#include <utility>

namespace engine
{

    Trade::Trade(
        std::string trade_id,
        std::string symbol,
        std::string taker_order_id,
        std::string maker_order_id,
        Side taker_side,
        double price,
        std::int64_t quantity)
        : trade_id_(std::move(trade_id)),
          symbol_(std::move(symbol)),
          taker_order_id_(std::move(taker_order_id)),
          maker_order_id_(std::move(maker_order_id)),
          taker_side_(taker_side),
          price_(price),
          quantity_(quantity)
    {
        if (trade_id_.empty())
        {
            throw std::invalid_argument(
                "trade id cannot be empty");
        }

        if (symbol_.empty())
        {
            throw std::invalid_argument(
                "trade symbol cannot be empty");
        }

        if (taker_order_id_.empty())
        {
            throw std::invalid_argument(
                "taker order id cannot be empty");
        }

        if (maker_order_id_.empty())
        {
            throw std::invalid_argument(
                "maker order id cannot be empty");
        }

        if (price_ <= 0.0)
        {
            throw std::invalid_argument(
                "trade price must be positive");
        }

        if (quantity_ <= 0)
        {
            throw std::invalid_argument(
                "trade quantity must be positive");
        }
    }

    const std::string &Trade::trade_id() const noexcept
    {
        return trade_id_;
    }

    const std::string &Trade::symbol() const noexcept
    {
        return symbol_;
    }

    const std::string &Trade::taker_order_id() const noexcept
    {
        return taker_order_id_;
    }

    const std::string &Trade::maker_order_id() const noexcept
    {
        return maker_order_id_;
    }

    Side Trade::taker_side() const noexcept
    {
        return taker_side_;
    }

    const std::string &Trade::buy_order_id() const noexcept
    {
        return taker_side_ == Side::BUY
                   ? taker_order_id_
                   : maker_order_id_;
    }

    const std::string &Trade::sell_order_id() const noexcept
    {
        return taker_side_ == Side::BUY
                   ? maker_order_id_
                   : taker_order_id_;
    }

    double Trade::price() const noexcept
    {
        return price_;
    }

    std::int64_t Trade::quantity() const noexcept
    {
        return quantity_;
    }

    bool Trade::is_valid() const noexcept
    {
        return !trade_id_.empty() && !symbol_.empty() && !taker_order_id_.empty() && !maker_order_id_.empty() && price_ > 0.0 && quantity_ > 0;
    }

} // namespace engine

File: analytics-py/src/analytics/models/trade.py
"""Trade domain model."""

from __future__ import annotations

from dataclasses import asdict, dataclass
from datetime import datetime
from decimal import Decimal
from typing import Any


@dataclass(frozen=True, slots=True)
class Trade:
    """An authoritative executed trade event from the engine."""

    event_id: str
    event_type: str
    trade_id: str
    symbol: str
    price: Decimal
    quantity: int
    timestamp: datetime
    taker_side: str
    buy_order_id: str | None = None
    sell_order_id: str | None = None
    taker_order_id: str | None = None
    maker_order_id: str | None = None

    def __post_init__(self) -> None:
        if not self.event_id:
            raise ValueError("event_id must not be empty")

        if self.event_type != "TRADE":
            raise ValueError("event_type must be TRADE")

        if not self.trade_id:
            raise ValueError("trade_id must not be empty")

        if not self.symbol:
            raise ValueError("symbol must not be empty")

        if self.price <= 0:
            raise ValueError("price must be positive")

        if self.quantity <= 0:
            raise ValueError("quantity must be positive")

        if self.taker_side not in ("BUY", "SELL"):
            raise ValueError("taker_side must be BUY or SELL")

    def to_dict(self) -> dict[str, Any]:
        """Convert the model to a JSON-friendly dictionary."""

        data = asdict(self)
        data["price"] = str(self.price)
        data["timestamp"] = self.timestamp.isoformat()
        return data
